### ISSUE TITLE
[Enhancement] Adding toString methods

### ISSUE THREAD

**Issue by briceambrosiak** [rithms/riot-api-java#71]
It would be nice if data objects had already overridden `toString()` methods displaying some key information. Like _(id, name, summonerLevel)_ for class `Summoner`, _(name, tier, queue)_ for class `League`, etc.

What do you think?
Thanks.

**Comment by rithms** [comment]
Sure, but for consistency sake we would have to add this for all the data objects. This leaves open the following questions:

1.) Which fields should be included/excluded for each DTO? Will all fields be included, or just fields deemed "important".
2.) What format should the information be printed in? (a list of values with no context, JSON, a line by line list of the field names and their values)

IMO a toString method should be a String representation of the object, but you mention using it to display only key information.

For example, when it comes to a toString() method for a League entry of a summoner, what is the expected output? Some may expect the output to be the JSON returned from the API, some may expect just the Tier of the summoner, some may expect Tier and Division, some may expect Tier, Division, League name, LP, etc.

If the method output doesn't match majority of developer expectations, they will just end up using the individual accessors for key information anyways, making the toString method useless.

I received an email a while ago from someone asking how he gets the name of a summoner because the toString() method wasn't working. I had to advise him to use the getName() method instead. This is an example of someone who would have different expectations than me.

IMO, if I were using a toString() method on a data object from this library, I would expect a JSON String with all the data returned by the Riot API. I don't understand why I would ever use that, but having a toString() method for objects would make the library more complete.

TL;DR: I'm down with having toString() methods, I'm just interested in hearing your opinion (and anyone elses) to the questions listed above.

**Comment by briceambrosiak** [comment]
Here's my point of view:

1. Only important field should be included. It would be then a matter of perspective to decide what is "important". But I'd say that the `toString()` method can be useful when logging and debugging your applications. You simply want to log your object to have some trace of it, and instead of having a useless memory pointer, you'd have some information.
The minimum "useful" information would be the technical id (of champion, league, etc.). Like I said, it's a matter of perspective and for a Summoner I'd put `(Id, Name)`.

2. No JSON format, as I think it's not the idea behind the Java toString() method. I like this notation : (val1; val2; val3), or (key1:val1; key2:val2; ...). JSON encoding should be delegate to transformation libraries that serializes DataObjects to JSON/XML formats.

If a developer want to access specific information, of course that he should use getters methods to retrieve it, and not toString(). It's just a matter of not being disappointed when you put your objects in some log and have nothing practical to use... ;)

**Comment by Linnun** [comment]
In my opinion there are only two ways for deciding what goes into toString():

1. Either make it minimalistic and only use some sort of 'primary keys' (for summoners this would be ID and name, for matches this would be the matchID, etc.)

2. Or make it complete and plot all fields. In some other projects I have toString() methods on some objects that print out every single field, much like this:

`UserObject {
	id: 1337,
	name: asd,
	level: 1
}`


I think I would prefer the second way for the following reason: When are you ever printing a DataObject except for when you're debugging? More information is always good when debugging.

**Comment by rithms** [comment]
Yeah, it doesn't have to be JSON. I was just using that as an example, since it has easy-to-read notation. Anything format that uses key-value pairs should be fine. I like and prefer @Linnun's second example for the same reasons given, but I also think there might be a happy medium between the two that might be useful for data objects with tons of not-so-useful fields.

**Comment by Linnun** [comment]
Would it be weird to have some kind of global 'verbose' flag which makes toString() methods output the complete data objects (all fields) while without that flag it would only return 'basics'?

**Comment by Linnun** [comment]
I started implementing such a functionality, but I stumbled upon difficulties with recursively printing a Dto and passing the verbosity flag on to childs when they are part of a list.



Example:

ChampionList contains a List<Champion>. Since List does not have a toString(boolean verbose) method, it will use the default toString() method of List, causing it to call the parameter-less toString() method of its items; so verbosity is not passed on to the Champion elements of the list.

While this is reasonably okay-ish for verbosity (you could say we don't want verbosity on each single list item), I also tried to pass on a `int depth` parameter, so that we would have some indentation to make it look pretty.

**Comment by Linnun** [comment]
I'm progressing. Here are some example outputs as of right now:

`PRINTING CHAMPION LIST
net.rithms.riot.api.endpoints.champion.dto.ChampionList {
   serialVersionUID: 987562252588662376
   champions: [266, 103, 84, 12, 32, 34, 1, 22, 136, 268, 432, 53, 63, 201, 51, 69, 31, 42, 122, 131, 119, 36, 245, 60, 28, 81, 9, 114, 105, 3, 41, 86, 150, 79, 104, 120, 74, 420, 39, 40, 59, 24, 126, 202, 222, 429, 43, 30, 38, 55, 10, 85, 121, 203, 240, 96, 7, 64, 89, 127, 236, 117, 99, 54, 90, 57, 11, 21, 62, 82, 25, 267, 75, 111, 76, 56, 20, 2, 61, 80, 78, 133, 33, 421, 58, 107, 92, 68, 13, 113, 35, 98, 102, 27, 14, 15, 72, 37, 16, 50, 134, 223, 163, 91, 44, 17, 412, 18, 48, 23, 4, 29, 77, 6, 110, 67, 45, 161, 254, 112, 8, 106, 19, 101, 5, 157, 83, 154, 238, 115, 26, 143]
}

PRINTING CHAMPION MASTERIES
[55: 18349 (Level 4), 223: 14290 (Level 4), 37: 11126 (Level 3), 201: 10249 (Level 3), 81: 9704 (Level 3), 3: 8806 (Level 3), 67: 6901 (Level 3), 16: 6520 (Level 3), 40: 5670 (Level 2), 15: 5663 (Level 2), 117: 5292 (Level 2), 99: 5232 (Level 2), 267: 4609 (Level 2), 85: 4487 (Level 2), 432: 4311 (Level 2), 25: 4124 (Level 2), 31: 3395 (Level 2), 134: 3164 (Level 2), 22: 3153 (Level 2), 245: 3135 (Level 2), 18: 2633 (Level 2), 21: 2426 (Level 2), 24: 1718 (Level 1), 78: 1668 (Level 1), 412: 1591 (Level 1), 17: 1576 (Level 1), 45: 1553 (Level 1), 82: 1436 (Level 1), 143: 1421 (Level 1), 102: 1419 (Level 1), 222: 1268 (Level 1), 10: 1112 (Level 1), 236: 1111 (Level 1), 83: 1037 (Level 1), 7: 1033 (Level 1), 112: 909 (Level 1), 39: 803 (Level 1), 61: 748 (Level 1), 20: 728 (Level 1), 9: 705 (Level 1), 86: 697 (Level 1), 51: 497 (Level 1), 203: 488 (Level 1), 8: 451 (Level 1), 202: 356 (Level 1), 150: 346 (Level 1), 119: 280 (Level 1), 50: 269 (Level 1), 161: 267 (Level 1), 107: 259 (Level 1), 429: 221 (Level 1), 80: 213 (Level 1), 92: 200 (Level 1), 89: 199 (Level 1), 127: 179 (Level 1), 12: 164 (Level 1), 69: 157 (Level 1), 420: 151 (Level 1), 268: 149 (Level 1)]

PRINTING CURRENT GAME
net.rithms.riot.api.endpoints.current_game.dto.CurrentGameInfo {
   serialVersionUID: -6094147804616587912
   bannedChampions: [1, 157, 105, 67, 245, 55]
   gameId: 2815409019
   gameLength: 186
   gameMode: CLASSIC
   gameQueueConfigId: 410
   gameStartTime: 1472298143430
   gameType: MATCHED_GAME
   mapId: 11
   observers: net.rithms.riot.api.endpoints.current_game.dto.Observer {
      serialVersionUID: 1866998515867420059
      encryptionKey: nPh1FS73XKoP58AsK9I3BzkM6mr9M/mQ
   }
   participants: [ILoveScuttlers, mbk10, Viatic, Yiimo, HCG Maajo SK, RosáŠchlüpfér, Maddox91, Robert1457, The Lâw, AxxidentProne]
   platformId: EUW1
   teamInfoOne: null
   teamInfoTwo: null
}

PRINTING FEATURED GAME
net.rithms.riot.api.endpoints.featured_game.dto.FeaturedGames {
   serialVersionUID: -8500784155234313042
   clientRefreshInterval: 300
   gameList: [2815398299 (CLASSIC), 2815399101 (CLASSIC), 2815389551 (CLASSIC), 2815389300 (CLASSIC), 2815388881 (CLASSIC)]
}
`

**Comment by briceambrosiak** [comment]
My opinion is still on the key attributes to choose to defined the `toString()`method. Verbosity and printing whole DTO as JSON could be another method like getJSON() or something like that. Keep it simple.

**Comment by Linnun** [comment]
This is only a preview of the verbose version.

**Comment by Linnun** [comment]
I have pushed a commit for this: https://github.com/rithms/riot-api-java/commit/d8ad7dfaf505c3d078aba2d0c781509987dcd709

I'll leave this issue open for now, please leave feedback.

**Comment by Linnun** [comment]
Closing now. Feel free to reopen if any changes need to be made.
